rp2040-hal/src/serial.rs

```diff
@@ -0,0 +1,50 @@
+//! Implementation for the embedded_hal::serial traits for the UART.
+// See [embedded-hal](https://docs.rs/embedded-hal/0.2.4/embedded_hal/serial/index.html) for more details
+
+use core::convert::Infallible;
+
+use crate::uart::{
+    UARTPeripheral,
+    UARTDevice,
+    Enabled
+};
+
+use embedded_hal::serial::{
+    Read,
+    Write
+};
+
+use nb::Error::WouldBlock;
+
+impl<D: UARTDevice> Read<u8> for UARTPeripheral<Enabled, D> {
+    type Error = Infallible;
+
+    fn read(&mut self) -> nb::Result<u8, Self::Error> {
```

Comment:
wouldn't this mean that you cannot read anything in that case instead of getting at least something?

Or error would contain in this case what it was able to read?

Reply:
Yes that is true, we could still return the data in the `Err()` variant.

Reply:
I'm not sure what you mean here @tdittr by the FIFO being overflowed and moreover how that state could have you wait too long between two calls to `read()`.
Still I feel like I have to disagree here. 
First, the `embedded-hal` traits' implementations "_Must erase device specific details._" (quoting the official documentation). So from that perspective, returning an error specific to the state of the FIFO doesn't stand. By the way, the FIFO might not be necessarily enabled : the default setting for the UART is to receive/send only one byte at a time.
Second, my understanding is that if the Receive FIFO is overflowed, that still means there's something to read. If there's something to read, then I feel like it would be an anti-pattern to return an error since `read()`'s goal is to read.

I feel like if there's an error to be had, that would be at the other side of the wire. If our Receive FIFO is full, then the sender's Transmit FIFO is going to overflow, and it's going to trigger a WouldBlock or equivalent on their side (and that's indeed what we do already).

Reply:
> I feel like if there's an error to be had, that would be at the other side of the wire. If our Receive FIFO is full, then the sender's Transmit FIFO is going to overflow, and it's going to trigger a WouldBlock or equivalent on their side (and that's indeed what we do already).

That would only hold in cases where flow control is being used. Without it either side can send data whenever it likes.

> I'm not sure what you mean here @tdittr by the FIFO being overflowed and moreover how that state could have you wait too long between two calls to read().

So if the other side sends data faster then we call `read()` at some point there will be no more space in the FIFO und thus the overrun error flag is set to high.

> Still I feel like I have to disagree here.
> First, the embedded-hal traits' implementations "Must erase device specific details." (quoting the official documentation). So from that perspective, returning an error specific to the state of the FIFO doesn't stand. 

Yes it has to erase device specific details so that a driver does not need to worry about *what* can go wrong, but it needs to be notified *that* something has gone wrong. That could also include other errors like parity errors. For example see the [`stm32f1xx_hal::serial::Error`](https://docs.rs/stm32f1xx-hal/0.7.0/stm32f1xx_hal/serial/enum.Error.html).

> By the way, the FIFO might not be necessarily enabled : the default setting for the UART is to receive/send only one byte at a time.

That does not actually change very much as then there would just be a FIFO of size 1. It will only overrun faster.

> Second, my understanding is that if the Receive FIFO is overflowed, that still means there's something to read. If there's something to read, then I feel like it would be an anti-pattern to return an error since read()'s goal is to read.

Yes that is a tough thing to decide, when to report the overrun. I would argue that it would be simpler to be reported as soon as the flag is set. Otherwise we need to keep the state of when the bit was set and count that down after any `read()` call.

Let me give an example. Assume the FIFO is turned off (size 1) and we have a simple protocol sending a coordinate like this:
```
X=-42
```

Now lets say the driver ready `X` and then is interrupted by a higher priority task. The other side sends `=` which is put into the FIFO. The driver has not yet resumed and the other side continues with `-` now there is no room in the FIFO so the `-` is dropped.

When the driver continues reading now it will read `=`. And continue with `42`. But now the `-` is lost which might be an important part of the message which now reads:
```
X=42
```

I hope that makes it more clear.

Reply:
Checked the datasheet again and it looks like without the FIFO activated it actually behaves like a shit-register. So the last character will be overritten by the new one.

So for my extreme example the driver should stop after reading `=` then `-` is received but overritten by `4`. :)

Reply:
Thanks for the detailed explanation, and I understand how an error would be required in case of overflow. I'll try to write something and you guys can tell me what you think !